compiler/rustc_middle/src/ty/fast_reject.rs

```diff
@@ -124,18 +118,23 @@ pub fn simplify_type(
         ty::Never => Some(NeverSimplifiedType),
         ty::Tuple(ref tys) => Some(TupleSimplifiedType(tys.len())),
         ty::FnPtr(ref f) => Some(FunctionSimplifiedType(f.skip_binder().inputs().len())),
-        ty::Projection(_) | ty::Param(_) => {
+        ty::Param(_) => {
             if can_simplify_params == SimplifyParams::Yes {
-                // In normalized types, projections don't unify with
-                // anything. when lazy normalization happens, this
-                // will change. It would still be nice to have a way
-                // to deal with known-not-to-unify-with-anything
-                // projections (e.g., the likes of <__S as Encoder>::Error).
                 Some(ParameterSimplifiedType)
             } else {
                 None
             }
         }
+        ty::Projection(_) => {
+            // Returning a simplified for projections can end up being unsound as it would
+            // stop us from considering not consider non blanket impls for `<_ as Trait>::Assoc` even
+            // though `_` can be inferred to a concrete type later at which point a concrete impl
+            // could actually apply.
+            //
+            // We could return `Some` here in cases where the input type is fully normalized, but
+            // I (@lcnr) don't know how to guarantee that.
+            None
```

Comment:
```suggestion
            // Returning a simplified for projections can end up being unsound. Consider this example:
            //
            // ```rust
            // trait Foo { }
            // impl Foo for u32 { }
            //
            // trait Bar { type Assoc; }
            // impl Bar for f32 { type Assoc = u32; }
            // impl Bar for f64 { type Assoc = u64; }
            // ```
            // If we are matching `<_ as Bar>::Assoc` against `Foo`, we might rule out the impl even though `_` could be inferred to `f32` and hence it might match.
            None
```